File: day11.go
package main

import (
	"strings"
)

const input = `hxbxwxba`

func increment(str string) string {
	init, last := str[:len(str)-1], str[len(str)-1]
	if last == 'z' {
		return increment(init) + "a"
	} else {
		return init + string(last+1)
	}
}

func rule1(str string) bool {
	for i := 0; i < len(str)-2; i++ {
		if str[i+1] == str[i]+1 && str[i+2] == str[i]+2 {
			return true
		}
	}
	return false
}

func rule2(str string) bool {
	return !strings.ContainsAny(str, "iol")
}

func rule3(str string) bool {
	alphabet := "abcdefghijklmnopqrstuvwxyz"
	n := 0
	for _, l := range alphabet {
		if strings.Contains(str, string(l)+string(l)) {
			n++
		}
	}
	return n >= 2
}

func valid(pw string) bool {
	return rule1(pw) && rule2(pw) && rule3(pw)
}

func nextPassword(pw string) string {
	for !valid(pw) {
		pw = increment(pw)
	}
	return pw
}

func main() {
	// part 1
	println(nextPassword(input))

	// part 2
	println(nextPassword(increment(nextPassword(input))))
}

File: 2015/day11.hs
import Data.List (group)

input :: String
input = "hxbxwxba"

increment :: String -> String
increment = reverse . inc . reverse where
	inc ('z':str) = 'a' : inc str
	inc (c:str)   = succ c : str

nextPassword :: String -> String
nextPassword = until valid increment where
	valid s = and [rule1 s, rule2 s, rule3 s]
	rule1 (x:y:z:s) = y == succ x && z == succ y || rule1 (y:z:s)
	rule1 _         = False
	rule2 s = not . or $ map (`elem` s) ['i','o','l']
	rule3 = (>= 2) . length . filter ((>= 2) . length) . group

part1 :: String -> String
part1 = nextPassword

part2 :: String -> String
part2 = nextPassword . increment . nextPassword

main :: IO ()
main = do
	print $ part1 input
	print $ part2 input
